zebROS_ws/src/behaviors/src/path_to_apriltag.cpp

```diff
@@ -0,0 +1,251 @@
+        sub_(nh_.subscribe<field_obj::Detection>("/tf_object_detection/tag_detection_world", 1, boost::bind(&PathToAprilTagAction::callback, this, _1))),
+        subImu_(nh_.subscribe<sensor_msgs::Imu>("/imu/zeroed_imu", 1, boost::bind(&PathToAprilTagAction::imuCallback, this, _1))),
+        timeout_(nh_params_.param<double>("timeout", timeout_, 0.2)) // roughly equal to apriltag detector fps
+    {
+        as_.start();
+    }
+
+    ~PathToAprilTagAction(void)
+    {
+    }
+
+    std::optional<tf2::Transform> getTransformToTag(uint32_t id, double rotation) {
+        if (latest_.objects.size() == 0 || ros::Time::now() - latest_.header.stamp > ros::Duration(timeout_)) {
+            return std::nullopt;
+        }
+        else {
+            for (auto object : latest_.objects) {
+                if (std::stoi(object.id) == id) {
+                    geometry_msgs::PointStamped point;
+                    point.point = object.location;
+                    point.header = latest_.header;
+                    geometry_msgs::PointStamped point_out;
+                    tf_buffer_.transform(point, point_out, "base_link");
+                    tf2::Transform tf;
+                    tf.setOrigin(tf2::Vector3(point_out.point.x, point_out.point.y, point_out.point.z));
+                    tf2::Quaternion q;
+                    q.setRPY(0, 0, (rotation - latestImuZ_));
+                    tf.setRotation(q);
+                    return std::optional<tf2::Transform>{tf};
+                }
+            }
+            return std::nullopt;
+        }
+    }
+
+    std::optional<trajectory_msgs::JointTrajectoryPoint> applyOffset(uint32_t id, const geometry_msgs::Pose &offset, double tagRotation) {
+        std::optional<tf2::Transform> transform_ = getTransformToTag(id, tagRotation);
+
+        if (!transform_.has_value()) {
+            ROS_INFO_STREAM("transform no value");
+            return std::nullopt;
+        }
+
+        tf2::Transform transform = transform_.value();
+
+        geometry_msgs::Transform gmt = tf2::toMsg(transform); // now tag -> base_link
+        geometry_msgs::TransformStamped gmts;
+        gmts.header.frame_id = "tag";
+        gmts.child_frame_id = "base_link";
+        gmts.transform = gmt;
+
+        // get offset, transform by inverse transform
+        geometry_msgs::Pose p;
+        tf2::doTransform(offset, p, gmts);
+        trajectory_msgs::JointTrajectoryPoint pt = generateTrajectoryPoint(p.position.x, p.position.y, getYaw(p.orientation));
+        return std::optional<trajectory_msgs::JointTrajectoryPoint>{pt};
+    }
+
+
+    void callback(const field_obj::DetectionConstPtr& msg) {
+        // need to use transforms...
+        // oh wait no we don't! we can tell spline srv what frame id it is relevant to
+        latest_ = *msg;
+    }
+
+    void imuCallback(const sensor_msgs::ImuConstPtr& msg) {
+        sensor_msgs::Imu msgNonPtr = *msg;
+        latestImuZ_ = getYaw(msgNonPtr.orientation);
```

Comment:
Just use msg->orientation?

Reply:
not sure how i missed that, very good point :)